service/matching/matcher.go

```diff
@@ -334,89 +320,70 @@ func (tm *TaskMatcher) Rate() float64 {
 	return tm.rateLimiter.Rate()
 }
 
-func (tm *TaskMatcher) pollOrForward(
-	ctx context.Context,
-	taskC <-chan *internalTask,
-	queryTaskC <-chan *internalTask,
-) (*internalTask, error) {
+func (tm *TaskMatcher) poll(ctx context.Context, queryOnly bool) (*internalTask, error) {
+	taskC, queryTaskC := tm.taskC, tm.queryTaskC
+	if queryOnly {
+		taskC = nil
+	}
+
+	// We want to effectively do a prioritized select, but Go select is random
+	// if multiple cases are ready, so split into multiple selects.
+	// The priority order is:
+	// 1. ctx.Done
+	// 2. taskC
+	// 3. queryTaskC
+	// 4. forwarding
```

Comment:
Would it be correct to say that now with this implementation we favor non-query tasks? Would it also be correct to say that with this code if there are enough non-query tasks that there is no guarantee that query tasks or forwarding work _ever_ execute?

Still coming up to speed here - sorry if I've missed things.

Reply:
Yeah, I mentioned that in the pr description. It may be an undesirable change.. someone with more experience should say.

I left it like that because this came from my change to prioritize sync over non-sync, which we decided to table for now.

Forwarding isn't more work, it's an extra mechanism to balance existing work, I think. In any case, I didn't change any semantics there: the local task channel(s) were already prioritized over forwarding, just with code that was hard to understand (duplicated and spread across multiple functions). I tried to just make it easier to understand.

Reply:
I think we want to prioritize query task because:
* query task is much rare compare to normal tasks.
* query is usually initiated by real person, normal task usually is not.